library/core/src/slice/sort.rs

```diff
@@ -198,9 +198,7 @@ where
             }
 
             // Choose the greater child.
-            if child + 1 < v.len() && is_less(&v[child], &v[child + 1]) {
-                child += 1;
-            }
+            child += (child + 1 < v.len() && is_less(&v[child], &v[child + 1])) as usize;
```

Comment:
Hmm, the `&&` here is still a branch, and a necessary one, so maybe there's an alternative way to write this to emphasize the parts?  Maybe something like
```suggestion
            if child + 1 < v.len() {
                // We need a branch to be sure not to out-of-bounds index,
                // but it's highly predictable.  The comparison, however,
                // is better done branchless, especially for primitives.
                child += is_less(&v[child], &v[child + 1]) as usize;
            }
```

Reply:
Great point, I really like your suggestion. All three generate different x86-64 assembly, I'll benchmark it tomorrow. Based on the assembly I suspect your suggestion will perform the best.

- Original https://godbolt.org/z/M9zrMWbW1
- Combined cond https://godbolt.org/z/7Es3avoYn
- Split cond https://godbolt.org/z/nhxoKcW3z

Reply:
Oh, nice, the `child += is_less(…, …)` version gives a `cmp`+`adc`.

Diff, in case it's helpful for anyone else, since I made it anyway: <https://godbolt.org/z/arcsxcEc5>.

Reply:
I just tested it and it drops it down to 249, neat.